Since 3.22.0CR1 fails ConfigProvider.getConfig() in the JUnit test (called from @MethodSource)

### Describe the bug

In Camel-quarkus's test AzureServiceBusTestResource (mocked one), with the upgrade to `3.22.0.CR1`, one test started to fail with the error:
```
[ERROR] Errors: 
[ERROR]   AzureServiceBusTest.produceConsumeOptions:359 » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
```
I was able to find the culprit. It is one of the last commits in 3.22.0.CR1 -  https://github.com/quarkusio/quarkus/commit/3c85fe4c847f73cc1588cb9d25b5fcf669d65dc7

I'm not sure, what exactly is causing the issue. But vefore the PR was merged, our test was working, with the PR merged, the test is not working.

Here is my forked branch, where the issue can be reproduced - https://github.com/JiriOndrusek/camel-quarkus/tree/quarkus-configprofile-error-reproducer




### Expected behavior

test should work with Quarkus 3.22.0CR1 (as was with the older version)

### Actual behavior

-  test is using methodSource - https://github.com/JiriOndrusek/camel-quarkus/blob/quarkus-configprofile-error-reproducer/integration-test-groups/azure/azure-servicebus/src/test/java/org/apache/camel/quarkus/component/azure/servicebus/it/AzureServiceBusTest.java#L355-L380
- from whic h the VonfigProvider.gtConfig is called - https://github.com/JiriOndrusek/camel-quarkus/blob/quarkus-configprofile-error-reproducer/integration-test-groups/azure/azure-servicebus/src/main/java/org/apache/camel/quarkus/component/azure/servicebus/it/AzureServiceBusHelper.java#L52
- this call throws an error. 

### How to Reproduce?

How to reproduce:

1. build Quarkus from the wrong commit, build camel-quarkus from my branch (CQ depends on quarkus 999-SNAPSHOT)
2. run the ~~`integration-tests/azure-grouped/`~~  `integration-test-groups/azure/azure-servicebus/` test module
3. you should see a failure in `produceConsumeMessage`
the test runs with mocked test resource, therefore there is no need to configure anything

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

Even more tests are failing -  can be simulated on the same branch as azure one:
```
2025-04-17T09:23:41.1389719Z [ERROR] Errors: 
2025-04-17T09:23:41.1391633Z [ERROR]   GroupedAwsSecretsManagerTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1395259Z [ERROR]   GroupedAws2CwTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1398455Z [ERROR]   GroupedAws2DdbTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1400369Z [ERROR]   GroupedAws2KinesisFirehoseTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1402335Z [ERROR]   GroupedAws2LambdaTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1403893Z [ERROR]   GroupedAws2S3Test.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1405758Z [ERROR]   GroupedAws2SqsSnsTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-04-17T09:23:41.1407348Z [ERROR]   GroupedAws2SqsTest.failingDefaultCredentialsProviderTest » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
```

I retested the scenario again:

- I built quarkus from `main` - commit 69e41f6ddbbf61b5a7112cb5fa7715cdd1e56856
- built camel-quarkus from my [branch](https://github.com/JiriOndrusek/camel-quarkus/tree/quarkus-configprofile-error-reproducer) - commit 87f8ee93a7010d9708b73ee9ab5587f4fc7761c6
- I started test by running `mvn clean test -f integration-test-groups/azure/azure-servicebus/` from camel-quarkus root dir.
-  the result is:
```
[INFO] 
[ERROR] Errors: 
[ERROR]   AzureServiceBusTest.produceConsumeOptions:359 » ServiceConfiguration io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
[INFO] 
[ERROR] Tests run: 6, Failures: 0, Errors: 1, Skipped: 2

```

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/MethodSourceTest.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/MethodSourceTest.java
@@ -8,6 +8,7 @@ import java.util.List;
 
 import jakarta.inject.Inject;
 
+import org.eclipse.microprofile.config.ConfigProvider;
 import org.hamcrest.CoreMatchers;
 import org.hamcrest.Matcher;
 import org.junit.jupiter.params.ParameterizedTest;
@@ -28,6 +29,9 @@ public class MethodSourceTest {
     public void testParameterResolver(UnusedBean.DummyInput dummyInput, Matcher<String> matcher) {
         UnusedBean.DummyResult dummyResult = unusedBean.dummy(dummyInput);
         assertThat(dummyResult.getResult(), matcher);
+
+        // Can we get config?
+        ConfigProvider.getConfig();
     }
 
     private static Collection<Arguments> provideDummyInput() {

--- a/integration-tests/maven/pom.xml
+++ b/integration-tests/maven/pom.xml
@@ -74,6 +74,14 @@
             <artifactId>quarkus-vertx-http-dev-ui-tests</artifactId>
             <scope>test</scope>
         </dependency>
+        <!-- This one is used by the tested projects so we need to create a dependency
+        to make sure GIB triggers the testing. --> 
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5</artifactId>
+            <scope>test</scope>
+            <version>${project.version}</version>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-project-core-extension-codestarts</artifactId>

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/ExecutionListener.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/ExecutionListener.java
@@ -25,14 +25,12 @@ public class ExecutionListener implements TestExecutionListener {
         Optional<TestSource> oSource = testIdentifier.getSource();
         if (oSource.isPresent()) {
             TestSource source = oSource.get();
-            if (source instanceof ClassSource) {
-                ClassSource cs = (ClassSource) source;
+            if (source instanceof ClassSource cs) {
                 ClassLoader classLoader = cs.getJavaClass().getClassLoader();
                 // Only adjust the TCCL in cases where we know the QuarkusTestExtension would be about to do it anyway
                 // We could check annotations, but that would be slow, and the assumption that only Quarkus Tests are loaded with the quarkus classloader should be a fair one
                 if (classLoader instanceof QuarkusClassLoader) {
-                    origCl = Thread.currentThread()
-                            .getContextClassLoader();
+                    origCl = Thread.currentThread().getContextClassLoader();
                     Thread.currentThread().setContextClassLoader(classLoader);
                 } else {
                     origCl = null;
@@ -43,9 +41,15 @@ public class ExecutionListener implements TestExecutionListener {
 
     @Override
     public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult result) {
-        if (origCl != null) {
-            // If execution is parallel this could be odd, but if execution is parallel any kind of TCCL manipulation will be ill-fated
-            Thread.currentThread().setContextClassLoader(origCl);
+        Optional<TestSource> oSource = testIdentifier.getSource();
+        if (oSource.isPresent()) {
+            TestSource source = oSource.get();
+            if (source instanceof ClassSource) {
+                if (origCl != null) {
+                    // If execution is parallel this could produce odd results, but if execution is parallel any kind of TCCL manipulation will be ill-fated
+                    Thread.currentThread().setContextClassLoader(origCl);
+                }
+            }
         }
     }
 }

